qwerty/ast_visitor/desugar.cpp

```diff
@@ -0,0 +1,36 @@
+#include "ast.hpp"
+#include "ast_visitor.hpp"
+
+// Rewrites basis literals of the form {a >> b, c >> d, ...} into {a, c, ...} >> {b, d, ...}
+bool DesugarVisitor::visit(ASTVisitContext &ctx, BasisLiteral &lit) {
+    for (auto const &elt : lit.elts) {
+        if (!dynamic_cast<BasisTranslation *>(elt.get())) {
+            // Forward this to type-checking so it can admit literals 
+            // of the form {'qbl1', 'qbl2', ...} or complain appropriately
+            return true;
+        }
+    }
+
+    std::vector<std::unique_ptr<ASTNode>> input_basis{};
+    input_basis.reserve(lit.elts.size());
+    std::vector<std::unique_ptr<ASTNode>> output_basis{};
+    output_basis.reserve(lit.elts.size());
+
+    for (auto const &elt : lit.elts) {
+        // Already know it's all BasisTranslation, static cast is safe.
+        BasisTranslation *btrans = static_cast<BasisTranslation *>(elt.get());
+        input_basis.push_back(std::move(btrans->basis_in));
+        output_basis.push_back(std::move(btrans->basis_out));
+    }
+
+    // Move everything everywhere everywhen
```

Comment:
I think you can remove this comment, unless it has more meaning than I'm realizing